samples/nrf9160/nrf_cloud_rest_fota/src/main.c

```diff
@@ -0,0 +1,821 @@
+
+		LOG_INF("---> Press button %d to enable FOTA in the device shadow", BTN_NUM);
+		LOG_INF("     Waiting %d seconds...", FOTA_ENABLE_WAIT_SEC);
+
+		ret = k_sem_take(&button_press_sem, K_SECONDS(FOTA_ENABLE_WAIT_SEC));
+
+		if (ret == 0) {
+			enable_fota_requested = true;
+		} else {
+			if (ret != -EAGAIN) {
+				LOG_ERR("k_sem_take error: %d", ret);
+			}
+			LOG_INF("Shadow will not be updated");
+		}
+	}
+
+	if (enable_fota_requested) {
+		LOG_INF("FOTA will be enabled after network connection is obtained");
+	}
+}
+
+static void do_fota_enable(void)
+{
+	int err;
+	/* Enable FOTA for bootloader, modem and application */
+	struct nrf_cloud_svc_info_fota fota = {
+		.bootloader = 1,
+		.modem = 1,
+		.application = 1
+	};
+	struct nrf_cloud_svc_info svc_inf = {
+		.fota = &fota,
+		.ui = NULL
+	};
+
+	err = generate_jwt();
+	if (err) {
+		return;
+	}
+
+	LOG_INF("Enabling FOTA...");
+	err = nrf_cloud_rest_shadow_service_info_update(&rest_ctx, device_id, &svc_inf);
+	if (err) {
+		LOG_ERR("Failed to enable FOTA, error: %d", err);
+	} else {
+		LOG_INF("FOTA enabled in device shadow");
+	}
+}
+
+static void process_pending_job(const int modem_lib_init_result)
+{
+	bool reboot_required = false;
+	int err;
+
+	if (pending_job.validate != FOTA_VALIDATE_PENDING) {
+		return;
+	}
+
+	if (pending_job.type == NRF_CLOUD_FOTA_MODEM) {
+
+		pending_job.validate = FOTA_VALIDATE_PASS;
+
+		switch (modem_lib_init_result) {
+		case MODEM_DFU_RESULT_OK:
+			LOG_INF("Modem FOTA update confirmed");
+			break;
+		case MODEM_DFU_RESULT_UUID_ERROR:
+		case MODEM_DFU_RESULT_AUTH_ERROR:
+		case MODEM_DFU_RESULT_HARDWARE_ERROR:
+		case MODEM_DFU_RESULT_INTERNAL_ERROR:
+			LOG_ERR("Modem FOTA error: %d", modem_lib_init_result);
+			pending_job.validate = FOTA_VALIDATE_FAIL;
+			break;
+		default:
+			LOG_INF("Modem FOTA result unknown: %d", modem_lib_init_result);
+			pending_job.validate = FOTA_VALIDATE_UNKNOWN;
+			break;
+		}
+
+		reboot_required = true;
+		LOG_INF("Modem FOTA complete, rebooting...");
+
+	} else if (pending_job.type == NRF_CLOUD_FOTA_APPLICATION ||
+		   pending_job.type == NRF_CLOUD_FOTA_BOOTLOADER) {
+
+		pending_job.validate = FOTA_VALIDATE_PASS;
+		fota_status = NRF_CLOUD_FOTA_SUCCEEDED;
+		fota_status_details = FOTA_STATUS_DETAILS_SUCCESS;
+
+		if (!boot_is_img_confirmed()) {
+			err = boot_write_img_confirmed();
+			if (err) {
+				LOG_ERR("App/boot FOTA update confirmation failed: %d",
+					err);
+				/* If this fails then MCUBOOT will revert
+				 * to the previous image on reboot
+				 */
+				pending_job.validate = FOTA_VALIDATE_FAIL;
+				reboot_required = true;
+				LOG_INF("Rebooting to revert to previous version...");
+			} else {
+				LOG_INF("App/boot FOTA update confirmed");
+			}
+		}
+	}
+
+	if (reboot_required) {
+		/* Sleep to display log message */
+		k_sleep(K_SECONDS(5));
+
+		/* Save validate status and reboot */
+		err = settings_save_one(FOTA_SETTINGS_FULL, &pending_job, sizeof(pending_job));
```

Comment:
Is this use of a settings area duplicated in other applications and samples which implement nRF Cloud FOTA?  What happens if this sample is used to FOTA a different program entirely?

Reply:
good point... the incoming FOTA app would need to also use this settings area. 
i will update the comments/docs to make this clear.
i think i will add a Kconfig option for a version # and/or string and then have conf/overlay so a user can build a test version for app FOTA.

this sample isn't meant to update from "nrf_cloud_rest_fota" to "asset_tracker_v2". 
it's just an example of how to implement FOTA in an application.

Reply:
changed my mind... updated to use the same settings area 😅

Reply:
So you're saying it is the same settings area as ATv1 / v2 use?  If so, great.

Reply:
Correct.  You can fota from this sample to atv1/2… and back.